### ISSUE TITLE
Add support for undefined number of items of form builder array 

### ISSUE THREAD

**Issue by natcohen** [angular/angular#12219]
[   ] bug report => search github for a similar issue or PR before submitting
[ x ] feature request
[   ] support request => Please do not submit support request here, instead see 

When using form builders in a reactive form, it is impossible to create an array of objects with an undefined number of objects. Let's say I have this form:

```
this.form = this.formBuilder.group({
    persons: this.formBuilder.array([
          this.formBuilder.group({
                name: [''],
                job: ['']
          })
    ]),
})
```

If my array is:

`var personsArray= [{name:'John', job:'plumber'}, {name:'Jack', job:'engineer'}, {name:'Bob', job:'cook'}];`

I can't really patch the form like this:

`this.form.patchValue({persons: personsArray})`

because it will only batch the first item of the array. There should be a way to iterate over the form array to create these controls through the `patchValue` function...

**Comment by kara** [comment]
Yeah, it would be nice to create controls that don’t yet exist if the current control list is too short. I want think a bit more about whether it should be part of the `patchValue` method, though. By design, it ignores controls it doesn’t recognize rather than creating them or throwing an error. This is mostly for the benefit of FormGroups, so super-sets of forms that contain properties like “id” and “created_at” don’t affect the form (we wouldn't want "id" to be added to the control list, for example).  We could either make the FormArray version behave differently to support this case or create a new method. Needs some thought.

In the meantime, it’s not impossible to achieve this yourself. If you only care about creating controls if your array value is longer than the current control list, you can add a while loop that creates any necessary controls and pushes them to the array:

```ts
while(this.form.get('persons').length < this.personsArray.length) {
   this.form.get('persons').push(this.buildPerson()); 
}
```

See full plunker example here:
http://plnkr.co/edit/cTHIwrIIimm74OKVAkB3?p=preview

**Comment by natcohen** [comment]
@kara Adding it to FormArray would be awesome! The temporary solution works but would add tons of those loops in very simple forms... I guess you have other priorities

**Comment by natcohen** [comment]
@kara would it be possible to also add support for unknown object (with deep iteration)? For example I have the following object: `{unknownProp1:{size:5, name:"pdf"}, unknownProp2:{size:4, name:"word"}}` and would like to have controls on size and name...

**Comment by nellyigiebor** [comment]
Hi,

I need help creating "reactive form builder array" from the below json data ( master --> detail wtih detail )

{
    "docNo": "abc0001",
    "transactionsDate": "01/01/2017",
    "storeId": "abcd",
    "invoiceDispatchDate": "01/01/2017",
    "deliveryAddress": "abc street, abc qtrs..",
    "purchaseDetail": [
        {
            "modelName": "abc leather",
            "price": "70$",
            "quantityDetail": [
                {
                    "colorDetail": [
                        {
                            "colorName": "Black",
                            "quantity": [
                                {
                                    "sizeNo": "11",
                                    "quantity": 12
                                },
                                {
                                    "sizeNo": "12",
                                    "quantity": 10
                                },
                                {
                                    "sizeNo": "10.5",
                                    "quantity": 200
                                }
                            ]
                        },
                        {
                            "colorName": "Red",
                            "quantity": [
                                {
                                    "sizeNo": "11.5",
                                    "quantity": 120
                                },
                                {
                                    "sizeNo": "11",
                                    "quantity": 50
                                },
                                {
                                    "sizeNo": "12",
                                    "quantity": 20
                                }
                            ]
                        }
                    ]
                }
            ]
        }
    ]
}

Thanks in advance.

Nelly

**Comment by Tiedye** [comment]
Any update on this?  It feels like missing functionality

**Comment by Tiedye** [comment]
I put together a component for my own project which fill in this gap (barely tested, just experimenting atm)

```ts
import { FormArray, AbstractControl, ValidatorFn, AbstractControlOptions, AsyncValidatorFn } from '@angular/forms';


export class FormArrayTemplate extends FormArray {
  constructor(
    public factory: (value: any, i: number) => AbstractControl,
    validatorOrOpts?: ValidatorFn | ValidatorFn[] | AbstractControlOptions | null,
    asyncValidator?: AsyncValidatorFn | AsyncValidatorFn[] | null
  ) {
    super([], validatorOrOpts, asyncValidator);
  }

  private _fixLength(value: any[]) {
    const length = value.length;
    if (this.length < length) {
      for (let i = this.length; i < length; ++i) {
        const control = this.factory(value[i], i);
        this.controls.push(control);
        (this as any)._registerControl(control);
      }
    } else if (this.length > length) {
      for (let i = this.length - 1; i >= length; --i) {
        (this.controls[i] as any)._registerOnCollectionChange(() => {});
      }
      this.controls.splice(length);
    }
  }

  patchValue(value: any[], options: {onlySelf?: boolean, emitEvent?: boolean} = {}) {
    this._fixLength(value);
    super.patchValue(value, options);
  }

  setValue(value: any[], options: {onlySelf?: boolean, emitEvent?: boolean} = {}) {
    this._fixLength(value);
    super.setValue(value, options);
  }

  reset(value: any = [], options: {onlySelf?: boolean, emitEvent?: boolean} = {}) {
    this._fixLength(value);
    super.reset(value, options);
  }
}
```
Since it extends `FormArray` it can be used anywhere a `FormArray` is expected (including in a FormBuilder)

**Comment by renanmontebelo** [comment]
This is incredibly difficult to implement without introducing some unexpected behaviors. Let's say you have this:

```
this.form = this.formBuilder.group({
    persons: this.formBuilder.array([
          this.formBuilder.group({
                name: ['', Validators.required], // note the validator!
                job: ['']
          })
    ]),
})
```

And then:

```
var personsArray= [{name:'John', job:'plumber'}, {name:'Jack', job:'engineer'}, {name:'Bob', job:'cook'}];
this.form.patchValue({persons: personsArray});
```

For sure John would have the validator, but what's the expectation about Jack and Bob? Should the validator be copied over from the 1st component to Jack and Bob too?

Let's imagine for a moment that yes, it should copy the validator from John to Jack and Bob. But now we can do something like this (on top of previous code):

```
this.form.persons.clear(); // remove all the array items;
this.form.patchValue({persons: personsArray});
```

Now there's no 1st item to copy from, and the validators are gone. I'm not even thinking about possible Observables so there might be more complications down the road.

I've been thinking about this feature for some time now. This and other FormArray edge-cases could be circumvented if we introduce some "schema" concept that you describe the structure and validators, and then bind a model to it. I was thinking of starting this as a separate 3rd party library but would be less generic than current Reactive Forms, so not sure if there are enough use cases to justify the effort.

**Comment by suppadeliux** [comment]
Is this still updated?.

**Comment by renanmontebelo** [comment]
> Is this still updated?.

There's no consensus (at least in this thread) as to how to implement this feature. I'd say that until someone drives this discussion with new ideas and gets consensus this feature is halted.
